### ISSUE TITLE
Linux build support

### ISSUE THREAD

**Issue by stackotter** [stackotter/swift-bundler#7]
A major goal of this project is to eventually support building apps for Linux. There are many competing file formats that can be used to bundle binaries and resources into a single executable/installable file for distribution. Two major ones are `deb` and `AppImage`.

To add Linux build support you'll have to create a `LinuxBundler` type implementing the `Bundler` protocol. The architecture of this project is very much based off the functional programming paradigm, so it may take a little getting used to writing relatively pure functions (in this case `static` methods) and using `Result` instead of `try` and `throw`. The functional style of programming generally has a lot of benefits for build systems and related tools.

I believe that the first format to be implemented should be `AppImage` as it's generally more cross platform than `deb`. However there are other formats such as `flatpak` (which take a bit more setup), so I'm definitely open to discuss the options further! Feel free to choose either `deb` or `AppImage` if you're more familiar with one or the other.

Modify [this line in `Bundler.swift`](https://github.com/stackotter/swift-bundler/blob/1ff7d96c1f3480225d589a844c81a7a4ecb8aee4/Sources/swift-bundler/Bundler/Bundler.swift#L30) to return `LinuxBundler.self` once you've created the `LinuxBundler` type, and you should be able to test out your implementation simply by building a simple test project (which can be created using `swift bundler create MyApp`).

As always, I'm very happy to help out contributors, do don't hesitate to contact me on Discord (@stackotter), Twitter ([@stackotter](https://twitter.com/stackotter), or via email (stackotter@stackotter.dev).

**Comment by tannerdsilva** [comment]
I might be able to move the ball forward on this. Certainly, it would be nice to have this for some projects I have on my horizon for next year.

Would you be opposed to me [building some of this source](https://github.com/AppImage/appimagetool/tree/main) into the solution? Or did you have another way in mind?

**Comment by stackotter** [comment]
> I might be able to move the ball forward on this.

Awesome!

> Would you be opposed to me [building some of this source](https://github.com/AppImage/appimagetool/tree/main) into the solution? Or did you have another way in mind?

That’s basically what I was planning too. I’d be fine if that tool was just considered a dependency for now (and we simply use Process to interface with it). But if you can turn the source into a form that can be built as a swiftpm library target, then all the better 👌

**Comment by stackotter** [comment]
Swift Bundler now supports Linux! 🎉 So far it can produce AppImages and rpms. I plan to tackle debs soon. For completeness, I've included a copy of my closing comment for #46 as it's probably more applicable to this thread anyway.

---

### Copy of #46 closing comment (with further elaboration)

I've finished off the AppImage bundler now, and introduced a `--bundler` command line option for specifying which bundler to use. I've also created an RPM bundler. The 3 bundlers available on Linux at the moment are `linuxAppImage` (the most portable), `linuxRPM` (useful when targeting Red Hat related distros), and `linuxGeneric` (useful during development). The only bundler available on macOS at the moment is `darwinApp`.

To learn more about each bundler see the documentation comments in the source code at [DarwinBundler.swift](https://github.com/stackotter/swift-bundler/blob/main/Sources/swift-bundler/Bundler/DarwinBundler.swift) (darwinApp), [GenericLinuxBundler.swift](https://github.com/stackotter/swift-bundler/blob/main/Sources/swift-bundler/Bundler/GenericLinuxBundler.swift) (linuxGeneric), [AppImageBundler.swift](https://github.com/stackotter/swift-bundler/blob/main/Sources/swift-bundler/Bundler/AppImageBundler.swift) (linuxAppImage), and [RPMBundler.swift](https://github.com/stackotter/swift-bundler/blob/main/Sources/swift-bundler/Bundler/RPMBundler.swift) (linuxRPM).

The default bundler on Linux is `linuxGeneric`. It provides a realistic runtime environment while avoiding the expensive compression and pre-processing steps that the distribution-focused bundlers must perform. Its output is a directory with a structure that roughly follows the [Filesystem Hierarchy Standard](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard). `linuxAppImage` and `linuxRPM` are both just thin wrappers around `linuxGeneric` that bundle up the output into a distributable file (an executable in `linuxAppImage`'s case, and an installer in `linuxRPM`'s case).

Bundling your app for distribution is simple with the `--bundler` option.

```sh
# Create an AppImage
swift bundler bundle --bundler linuxAppImage

# Create an RPM
swift bundler bundle --bundler linuxRPM
```
